From 85780a6e6c554469c3077040c52b5f7ee842c507 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Wed, 30 Mar 2011 07:10:48 -0600
Subject: [PATCH 032/158] - move adding fop to the fol into reqh handle
 function. This is an clearly generic action that allows to save many lines of
 code in fo_state()

---
 ioservice/io_foms.c |  12 +--
 mdservice/md_foms.c | 279 ++++++++++++++++++++++++----------------------------
 reqh/reqh.c         |   8 ++
 3 files changed, 135 insertions(+), 164 deletions(-)

diff --git a/ioservice/io_foms.c b/ioservice/io_foms.c
index 35958f8..35b9d9c 100644
--- a/ioservice/io_foms.c
+++ b/ioservice/io_foms.c
@@ -216,16 +216,6 @@ int c2_fom_cob_rwv_state(struct c2_fom *fom)
 	        site->s_datastore->ds_dom, ctx->ft_tx);
 	C2_ASSERT(result == 0);
 
-	if(fom_obj->fcrw_fop->f_type->ft_code == c2_io_service_writev_opcode)
-	{
-		/* 
-		 * Make an FOL transaction record.
-		 */
-		result = c2_fop_fol_rec_add(fom_obj->fcrw_fop, 
-				            ctx->ft_fol, &ctx->ft_tx->tx_dbtx);
-		C2_ASSERT(result == 0);
-	}
-
 	/*
 	 * Allocate and find out the c2_stob object from given domain. 
 	 */
@@ -376,7 +366,7 @@ int c2_dummy_req_handler(struct c2_service *s, struct c2_fop *fop,
 			 struct c2_stob_domain *dom)
 {
 	struct c2_fop_ctx 	ctx;
-	int			result = 0;
+	int			result;
 	struct c2_fom	       *fom = NULL;
 
 	ctx.ft_service = s;
diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index ba0309e..b207192 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -90,23 +90,20 @@ static int c2_md_fom_create_state(struct c2_fom *fom)
         site = ctx->ft_site;
         C2_ASSERT(site != NULL);
 
-        rc = c2_fop_fol_rec_add(fop, ctx->ft_fol, &ctx->ft_tx->tx_dbtx);
-        if (rc == 0) {
-                rc = c2_md_store_alloc(site->s_mdstore, &cob);
-                if (rc)
-                        goto out;
-
-                req = c2_fop_data(fop);
-                rep = c2_fop_data(fop_rep);
-
-                rc = c2_md_create_permissions(req, ctx);
-                if (rc) {
-                        c2_cob_put(cob);
-                        goto out;
-                }
-                rc = c2_md_store_create(site->s_mdstore, cob, req, rep, ctx);
+        rc = c2_md_store_alloc(site->s_mdstore, &cob);
+        if (rc)
+                goto out;
+
+        req = c2_fop_data(fop);
+        rep = c2_fop_data(fop_rep);
+
+        rc = c2_md_create_permissions(req, ctx);
+        if (rc) {
                 c2_cob_put(cob);
+                goto out;
         }
+        rc = c2_md_store_create(site->s_mdstore, cob, req, rep, ctx);
+        c2_cob_put(cob);
 out:
 	fom->fo_phase = rc ? FOPH_FAILED : FOPH_DONE;
 	c2_md_req_fom_fini(fom);
@@ -147,28 +144,25 @@ static int c2_md_fom_link_state(struct c2_fom *fom)
         site = ctx->ft_site;
         C2_ASSERT(site != NULL);
 
-        rc = c2_fop_fol_rec_add(fop, ctx->ft_fol, &ctx->ft_tx->tx_dbtx);
-        if (rc == 0) {
-                req = c2_fop_data(fop);
-                body = &req->l_body;
+        req = c2_fop_data(fop);
+        body = &req->l_body;
 
-                rc = c2_md_link_permissions(req, ctx);
-                if (rc)
-                        goto out;
+        rc = c2_md_link_permissions(req, ctx);
+        if (rc)
+                goto out;
 
-                c2_md_make_fid(&fid, &body->b_tfid);
+        c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
-                                        C2_MD_STORE_LOCATE_STORE,
-                                        &ctx->ft_tx->tx_dbtx);
-                if (rc)
-                        goto out;
+        rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                C2_MD_STORE_LOCATE_STORE,
+                                &ctx->ft_tx->tx_dbtx);
+        if (rc)
+                goto out;
 
-                rep = c2_fop_data(fop_rep);
+        rep = c2_fop_data(fop_rep);
 
-                rc = c2_md_store_link(site->s_mdstore, cob, req, rep, ctx);
-                c2_cob_put(cob);
-        }
+        rc = c2_md_store_link(site->s_mdstore, cob, req, rep, ctx);
+        c2_cob_put(cob);
 out:
 	fom->fo_phase = rc ? FOPH_FAILED : FOPH_DONE;
 	c2_md_req_fom_fini(fom);
@@ -209,28 +203,25 @@ static int c2_md_fom_unlink_state(struct c2_fom *fom)
         site = ctx->ft_site;
         C2_ASSERT(site != NULL);
 
-        rc = c2_fop_fol_rec_add(fop, ctx->ft_fol, &ctx->ft_tx->tx_dbtx);
-        if (rc == 0) {
-                req = c2_fop_data(fop);
-                body = &req->u_body;
+        req = c2_fop_data(fop);
+        body = &req->u_body;
 
-                rc = c2_md_unlink_permissions(req, ctx);
-                if (rc)
-                        goto out;
+        rc = c2_md_unlink_permissions(req, ctx);
+        if (rc)
+                goto out;
 
-                c2_md_make_fid(&fid, &body->b_tfid);
+        c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
-                                        C2_MD_STORE_LOCATE_STORE,
-                                        &ctx->ft_tx->tx_dbtx);
-                if (rc)
-                        goto out;
+        rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                C2_MD_STORE_LOCATE_STORE,
+                                &ctx->ft_tx->tx_dbtx);
+        if (rc)
+                goto out;
 
-                rep = c2_fop_data(fop_rep);
+        rep = c2_fop_data(fop_rep);
 
-                rc = c2_md_store_unlink(site->s_mdstore, cob, req, rep, ctx);
-                c2_cob_put(cob);
-        }
+        rc = c2_md_store_unlink(site->s_mdstore, cob, req, rep, ctx);
+        c2_cob_put(cob);
 out:
 	fom->fo_phase = rc ? FOPH_FAILED : FOPH_DONE;
 	c2_md_req_fom_fini(fom);
@@ -271,28 +262,25 @@ static int c2_md_fom_rename_state(struct c2_fom *fom)
         site = ctx->ft_site;
         C2_ASSERT(site != NULL);
 
-        rc = c2_fop_fol_rec_add(fop, ctx->ft_fol, &ctx->ft_tx->tx_dbtx);
-        if (rc == 0) {
-                req = c2_fop_data(fop);
-                body = &req->r_sbody;
+        req = c2_fop_data(fop);
+        body = &req->r_sbody;
 
-                rc = c2_md_rename_permissions(req, ctx);
-                if (rc)
-                        goto out;
+        rc = c2_md_rename_permissions(req, ctx);
+        if (rc)
+                goto out;
 
-                c2_md_make_fid(&fid, &body->b_tfid);
+        c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
-                                        C2_MD_STORE_LOCATE_STORE,
-                                        &ctx->ft_tx->tx_dbtx);
-                if (rc)
-                        goto out;
+        rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                C2_MD_STORE_LOCATE_STORE,
+                                &ctx->ft_tx->tx_dbtx);
+        if (rc)
+                goto out;
 
-                rep = c2_fop_data(fop_rep);
+        rep = c2_fop_data(fop_rep);
 
-                rc = c2_md_store_rename(site->s_mdstore, cob, req, rep, ctx);
-                c2_cob_put(cob);
-        }
+        rc = c2_md_store_rename(site->s_mdstore, cob, req, rep, ctx);
+        c2_cob_put(cob);
 out:
 	fom->fo_phase = rc ? FOPH_FAILED : FOPH_DONE;
 	c2_md_req_fom_fini(fom);
@@ -333,33 +321,30 @@ static int c2_md_fom_open_state(struct c2_fom *fom)
         site = ctx->ft_site;
         C2_ASSERT(site != NULL);
 
-        rc = c2_fop_fol_rec_add(fop, ctx->ft_fol, &ctx->ft_tx->tx_dbtx);
-        if (rc == 0) {
-                req = c2_fop_data(fop);
-                body = &req->o_body;
+        req = c2_fop_data(fop);
+        body = &req->o_body;
 
-                rc = c2_md_open_permissions(req, ctx);
-                if (rc)
-                        goto out;
+        rc = c2_md_open_permissions(req, ctx);
+        if (rc)
+                goto out;
 
-                c2_md_make_fid(&fid, &body->b_tfid);
+        c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
-                                        C2_MD_STORE_LOCATE_STORE,
-                                        &ctx->ft_tx->tx_dbtx);
-                if (rc == -ENOENT) {
-                        /*
-                         * @todo: create file if open mode says so.
-                         */
-                } else if (rc) {
-                        goto out;
-                }
+        rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                C2_MD_STORE_LOCATE_STORE,
+                                &ctx->ft_tx->tx_dbtx);
+        if (rc == -ENOENT) {
+                /*
+                 * @todo: create file if open mode says so.
+                 */
+        } else if (rc) {
+                goto out;
+        }
 
-                rep = c2_fop_data(fop_rep);
+        rep = c2_fop_data(fop_rep);
 
-                rc = c2_md_store_open(site->s_mdstore, cob, req, rep, ctx);
-                c2_cob_put(cob);
-        }
+        rc = c2_md_store_open(site->s_mdstore, cob, req, rep, ctx);
+        c2_cob_put(cob);
 out:
 	fom->fo_phase = rc ? FOPH_FAILED : FOPH_DONE;
 	c2_md_req_fom_fini(fom);
@@ -394,23 +379,20 @@ static int c2_md_fom_close_state(struct c2_fom *fom)
         site = ctx->ft_site;
         C2_ASSERT(site != NULL);
 
-        rc = c2_fop_fol_rec_add(fop, ctx->ft_fol, &ctx->ft_tx->tx_dbtx);
-        if (rc == 0) {
-                req = c2_fop_data(fop);
-                body = &req->c_body;
+        req = c2_fop_data(fop);
+        body = &req->c_body;
 
-                c2_md_make_fid(&fid, &body->b_tfid);
+        c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
-                                        C2_MD_STORE_LOCATE_OPEN,
-                                        &ctx->ft_tx->tx_dbtx);
-                if (rc)
-                        goto out;
-                rep = c2_fop_data(fop_rep);
+        rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                C2_MD_STORE_LOCATE_OPEN,
+                                &ctx->ft_tx->tx_dbtx);
+        if (rc)
+                goto out;
+        rep = c2_fop_data(fop_rep);
 
-                rc = c2_md_store_close(site->s_mdstore, cob, req, rep, ctx);
-                c2_cob_put(cob);
-        }
+        rc = c2_md_store_close(site->s_mdstore, cob, req, rep, ctx);
+        c2_cob_put(cob);
 out:
 	fom->fo_phase = rc ? FOPH_FAILED : FOPH_DONE;
 	c2_md_req_fom_fini(fom);
@@ -451,28 +433,25 @@ static int c2_md_fom_setattr_state(struct c2_fom *fom)
         site = ctx->ft_site;
         C2_ASSERT(site != NULL);
 
-        rc = c2_fop_fol_rec_add(fop, ctx->ft_fol, &ctx->ft_tx->tx_dbtx);
-        if (rc == 0) {
-                req = c2_fop_data(fop);
-                body = &req->s_body;
+        req = c2_fop_data(fop);
+        body = &req->s_body;
 
-                rc = c2_md_setattr_permissions(req, ctx);
-                if (rc)
-                        goto out;
+        rc = c2_md_setattr_permissions(req, ctx);
+        if (rc)
+                goto out;
 
-                c2_md_make_fid(&fid, &body->b_tfid);
+        c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
-                                        C2_MD_STORE_LOCATE_STORE,
-                                        &ctx->ft_tx->tx_dbtx);
-                if (rc)
-                        goto out;
+        rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                C2_MD_STORE_LOCATE_STORE,
+                                &ctx->ft_tx->tx_dbtx);
+        if (rc)
+                goto out;
 
-                rep = c2_fop_data(fop_rep);
+        rep = c2_fop_data(fop_rep);
 
-                rc = c2_md_store_setattr(site->s_mdstore, cob, req, rep, ctx);
-                c2_cob_put(cob);
-        }
+        rc = c2_md_store_setattr(site->s_mdstore, cob, req, rep, ctx);
+        c2_cob_put(cob);
 out:
 	fom->fo_phase = rc ? FOPH_FAILED : FOPH_DONE;
 	c2_md_req_fom_fini(fom);
@@ -513,28 +492,25 @@ static int c2_md_fom_getattr_state(struct c2_fom *fom)
         site = ctx->ft_site;
         C2_ASSERT(site != NULL);
 
-        rc = c2_fop_fol_rec_add(fop, ctx->ft_fol, &ctx->ft_tx->tx_dbtx);
-        if (rc == 0) {
-                req = c2_fop_data(fop);
-                body = &req->g_body;
+        req = c2_fop_data(fop);
+        body = &req->g_body;
 
-                rc = c2_md_getattr_permissions(req, ctx);
-                if (rc)
-                        goto out;
+        rc = c2_md_getattr_permissions(req, ctx);
+        if (rc)
+                goto out;
 
-                c2_md_make_fid(&fid, &body->b_tfid);
+        c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
-                                        C2_MD_STORE_LOCATE_STORE,
-                                        &ctx->ft_tx->tx_dbtx);
-                if (rc)
-                        goto out;
+        rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                C2_MD_STORE_LOCATE_STORE,
+                                &ctx->ft_tx->tx_dbtx);
+        if (rc)
+                goto out;
 
-                rep = c2_fop_data(fop_rep);
+        rep = c2_fop_data(fop_rep);
 
-                rc = c2_md_store_getattr(site->s_mdstore, cob, req, rep, ctx);
-                c2_cob_put(cob);
-        }
+        rc = c2_md_store_getattr(site->s_mdstore, cob, req, rep, ctx);
+        c2_cob_put(cob);
 out:
 	fom->fo_phase = rc ? FOPH_FAILED : FOPH_DONE;
 	c2_md_req_fom_fini(fom);
@@ -575,28 +551,25 @@ static int c2_md_fom_readdir_state(struct c2_fom *fom)
         site = ctx->ft_site;
         C2_ASSERT(site != NULL);
 
-        rc = c2_fop_fol_rec_add(fop, ctx->ft_fol, &ctx->ft_tx->tx_dbtx);
-        if (rc == 0) {
-                req = c2_fop_data(fop);
-                body = &req->r_body;
+        req = c2_fop_data(fop);
+        body = &req->r_body;
 
-                rc = c2_md_readdir_permissions(req, ctx);
-                if (rc)
-                        goto out;
+        rc = c2_md_readdir_permissions(req, ctx);
+        if (rc)
+                goto out;
 
-                c2_md_make_fid(&fid, &body->b_tfid);
+        c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
-                                        C2_MD_STORE_LOCATE_STORE,
-                                        &ctx->ft_tx->tx_dbtx);
-                if (rc)
-                        goto out;
+        rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                C2_MD_STORE_LOCATE_STORE,
+                                &ctx->ft_tx->tx_dbtx);
+        if (rc)
+                goto out;
 
-                rep = c2_fop_data(fop_rep);
+        rep = c2_fop_data(fop_rep);
 
-                rc = c2_md_store_readdir(site->s_mdstore, cob, req, rep, ctx);
-                c2_cob_put(cob);
-        }
+        rc = c2_md_store_readdir(site->s_mdstore, cob, req, rep, ctx);
+        c2_cob_put(cob);
 out:
 	fom->fo_phase = rc ? FOPH_FAILED : FOPH_DONE;
 	c2_md_req_fom_fini(fom);
diff --git a/reqh/reqh.c b/reqh/reqh.c
index f3ad41c..e52b058 100644
--- a/reqh/reqh.c
+++ b/reqh/reqh.c
@@ -50,9 +50,17 @@ int c2_reqh_fop_handle(struct c2_reqh *reqh, struct c2_fop *fop)
         ctx.ft_fol  = reqh->rh_fol;
         ctx.ft_env = &env;
         ctx.ft_tx = &tx;
+
 //	ctx.ft_service = s;
 //	ctx.fc_cookie  = cookie;
 
+        /*
+         * Add this fop to the fol.
+         */
+        result = c2_fop_fol_rec_add(fop, ctx.ft_fol, &tx.tx_dbtx);
+        if (result)
+                goto out;
+
 	/*
 	 * Reqh generic phases will be run here that will do 
 	 * the standard actions like authentication, authorization,
-- 
1.8.3.2

